syntax = "proto3";

package indigo.net.pb;

/*************************************************\
 *   Sanctify game networking protocol buffers   *
 *                                               *
 * Below is an enumeration of possible messages  *
 *  that are sent between the client and server  *
 *  in a Sanctify game match                     *
\*************************************************/

/**
 * Initial connection message - use to authenticate the connecting client as a certain user.
 */
message InitialConnectionRequest {
  // Token that both authenticates the player and serves and identifies the game
  //  which they have been given permission to join. Check against an external service
  //  usually.
  string player_token = 1;
}

message InitialConnectionResponse {
  enum ResponseType {
    UNKNOWN = 0;
    ACCEPTED = 1;
    INVALID_TOKEN = 2;
    TIMEOUT = 3;
    BAD_FORMAT = 4;
  }

  ResponseType response_type = 1;

  // TODO (sessamekesh): enumerate which resources the client should preload
}

/*********************************************************
 * Union of all messages that a server can send to a client
 *********************************************************/
message GameServerMessage {
  oneof msg_body {
    InitialConnectionResponse initial_connection_response = 1;
  }
}

/***********************************************************
 * Union of all messages that a client can send to a server
 **********************************************************/
message GameClientMessage {
  oneof msg_body {
    InitialConnectionRequest initial_connection_request = 1;
  }
}